versions/1.1/SPEC.md

```diff
@@ -0,0 +1,5775 @@
+
+##### Array[X]
+
+An `Array` represents an ordered list of elements that are all of the same type.
+
+An array value can be declared with comma-separated values in brackets (`[]`), and a specific (zero-based) index of an `Array` can be accessed by placing the index in brackets after the declaration name. Accessing a non-existent index of an `Array` results in an error.
+
+```wdl
+Array[File] files = ["/path/to/file1", "/path/to/file2"]
+File f = files[0]
+
+Array[Int] empty = []
+empty[0]  # error!
+```
+
+An `Array` may be empty, unless it is declared using the non-empty postfix quantifier `+`, in which case it must contain at least one element. Note that the `+` and `?` postfix quantifiers can be combined to declare an array that is either undefined or non-empty. Attempting to assign an empty array literal to a non-empty `Array` declaration results in an error. Otherwise, the non-empty assertion is only checked at runtime: binding an empty array to an `Array[T]+` input or function argument is a runtime error. 
+
+```wdl
+# array that must contain at least one Float
+Array[Float]+ nonempty = [0.0]
+# array that must contain at least one Int?
+# (which may have an undefined value)
+Array[Int?]+ nonempty2 = [None, 1]
+# array that can be undefined or must contain
+# at least one Int
+Array[Int]+? nonempty4
+Array[Int]+? nonempty5 = [0.0]
+
+Array[Boolean]+ nonempty3 = []  # error!
+Array[Int]+? nonempty6 = []     # error!
+```
+
+##### Pair[X, Y]
+
+A `Pair` represents two associated values, which may be of different types In other programming languages, a `Pair` might be called a "two-tuple".
+
+A `Pair` can be declared with comma-separated values in parentheses (`()`). The components of a pair value are accessed using its `left` and `right` accessors.
+
+```wdl
+Pair[Int, Array[String]] data = (5, ["hello", "goodbye"])
+Int i = p.left
+String hello = data.right[0]
+```
+
+##### Map[P, Y]
+
+A `Map` represents an associative array of key-value pairs. All of the keys must be of the same (primitive) type, and all of the values must be of the same type, but keys and values can be different types.
+
+A `Map` can be declared with comma-separated key-value pairs in braces (`{}`). Key-value pairs are delimited by `:`. The value of a specific key can be accessed by placing the key in brackets after the declaration name. Accessing a non-existent key of a `Map` results in an error.
+
+```wdl
+Map[Int, Int] int_to_int = {1: 10, 2: 11}
+Map[String, Int] string_to_int = { "a": 1, "b": 2 }
+Map[File, Array[Int]] file_to_ints = {
+  "/path/to/file1": [0, 1, 2],
+  "/path/to/file2": [9, 8, 7]
+}
+
+string_to_int["b"]  # => 2
+string_to_int["c"]  # error!  
+```
+
+A `Map` is insertion ordered, meaning the order in which elements are added to the `Map` is preserved, for example when [✨ converting a `Map` to an array of `Pair`s](#-arraypairp-y-as_pairsmapp-y).
+
+```wdl
```

Comment:
These types of examples might be more clear as well if you did something like this:
```wdl
## Variable definition:
Map[Int, String] int_to_string = { 2: "hello", 1: "goodbye" }

## Variable evaluation at runtime:
as_pairs(int_to_string)  # => [(2, "hello"), (1, "goodbye")]
```
Because technically you wouldn't just WRITE the evaluation statement in the WDL in the same section as the variable definition.  Might help folks understand the timing of when to write what sort of format.  Also you haven't introduced what `as_pairs` actually is, b/c the library section hasn't been read yet. ;) I think that's sort of fine, but call out that that's a runtime thing.

Reply:
Right - these need to be cleaned up throughout the whole spec. I'll work on that.

Reply:
done